File: models/Gpt-3.5-turbo/py/code/advanced_techniques/Reachable_Nodes.py

def dfs(node, graph, visited, reachable):
    visited[node] = True
    reachable[node] += 1

    for neighbor in graph[node]:
        if not visited[neighbor]:
            dfs(neighbor, graph, visited, reachable)

n, m = map(int, input().split())

graph = {i: [] for i in range(1, n+1)}

for _ in range(m):
    a, b = map(int, input().split())
    graph[a].append(b)

reachable = [0] * (n+1)
visited = [False] * (n+1)

for node in range(1, n+1):
    if not visited[node]:
        dfs(node, graph, visited, reachable)

print(' '.join(map(str, reachable[1:])))

File: models/Gpt-3.5-turbo/cpp/code/problems/advanced_techniques/Reachable_Nodes.cpp

#include <iostream>
#include <vector>

using namespace std;

vector<vector<int>> adj;
vector<bool> visited;
vector<int> reachable;

void dfs(int node) {
    visited[node] = true;
    reachable[node]++;

    for (int neighbor : adj[node]) {
        if (!visited[neighbor]) {
            dfs(neighbor);
        }
    }
}

int main() {
    int n, m;
    cin >> n >> m;

    adj.resize(n + 1);
    visited.resize(n + 1);
    reachable.resize(n + 1);

    for (int i = 0; i < m; i++) {
        int a, b;
        cin >> a >> b;
        adj[a].push_back(b);
    }

    for (int i = 1; i <= n; i++) {
        if (!visited[i]) {
            dfs(i);
        }
    }

    for (int i = 1; i <= n; i++) {
        cout << reachable[i] << " ";
    }

    return 0;
}
